src/engine/Union.cpp

```diff
@@ -233,4 +237,73 @@ void Union::computeUnion(
     writeColumn(left, targetColumn, leftCol, 0u);
     writeColumn(right, targetColumn, rightCol, left.size());
   }
+  return res;
+}
+
+// _____________________________________________________________________________
+template <bool left>
+std::vector<size_t> Union::computePermutation() const {
+  size_t startOfUndefColumns = _subtrees[left ? 0 : 1]->getResultWidth();
+  std::vector<size_t> permutation(_columnOrigins.size());
+  for (size_t targetColIdx = 0; targetColIdx < _columnOrigins.size();
+       ++targetColIdx) {
+    size_t originIndex = _columnOrigins.at(targetColIdx)[left ? 0 : 1];
+    if (originIndex == NO_COLUMN) {
+      originIndex = startOfUndefColumns;
+      startOfUndefColumns++;
+    }
+    permutation[targetColIdx] = originIndex;
+  }
+  return permutation;
```

Comment:
In retrospective I'm not quite sure if there's an actual performance benefit to compute the permutation once and reuse it for all id tables or if it would be possible to just inline it into the algorithm that does the actual swapping, saving the memory allocation.

Reply:
Maybe it would be better to just store pairs of indices of swap operations?

Reply:
It is perfectly fine as is, the assumption always is, that columns and blocks are sufficiently large, s.t. you can do something mildly expensive for each of them.